pkg/destroy/vsphere/vsphere.go

```diff
@@ -227,59 +244,92 @@ func (o *ClusterUninstaller) Run() (*installertypes.ClusterQuota, error) {
 		return nil, errors.Errorf("Expected 1 Folder per tag but got %d", len(folderList))
 	}
 
-	if len(virtualMachineList) > 0 {
-		o.Logger.Debug("Find VirtualMachine objects")
-		virtualMachineMoList, err := getVirtualMachineManagedObjects(context.TODO(), o.Client, virtualMachineList)
-		if err != nil {
-			return nil, err
-		}
-		o.Logger.Debug("Delete VirtualMachines")
-		err = deleteVirtualMachines(context.TODO(), o.Client, virtualMachineMoList, o.Logger)
-		if err != nil {
-			return nil, err
-		}
-	} else {
+	if len(virtualMachineList) == 0 {
 		o.Logger.Debug("No VirtualMachines found")
 	}
 
-	if len(folderList) > 0 {
-		o.Logger.Debug("Find Folder objects")
-		folderMoList, err := getFolderManagedObjects(context.TODO(), o.Client, folderList)
-		if err != nil {
-			o.Logger.Errorln(err)
-			return nil, err
-		}
+	// First, make sure all the VMs are terminated so the cluster is not creating any new resources
+	var virtualMachineMoList []mo.VirtualMachine
+	err = wait.PollInfinite(
+		time.Second*10,
+		func() (done bool, err error) {
+			if len(virtualMachineList) > 0 && len(virtualMachineMoList) == 0 {
+				o.Logger.Debug("Find VirtualMachine objects")
+				virtualMachineMoList, err = getVirtualMachineManagedObjects(context.TODO(), o.Client, virtualMachineList)
```

Comment:
Let's get the list anew each time through. We want to loop until the machine is reporting powered off. This requires re-querying for the machine each time. We also want a separate list for machines that we have already sent the power-off request for, so that we are not repeating that request.

Moreover, I think we need to actually re-query for object by tag each time through this loop. Until all of the machines are powered-off, the cluster may attempt to create new machines. We should be resilient to that.